eclipse-che/che-dashboard
# feat: implement more specific error pages for 'workspace stopped' err…

…or page

Signed-off-by: David Kwon <dakwon@redhat.com>



### What does this PR do?

Currently there is a generic error page that is displayed when trying to access a workspace url (a devworkspace's `status.mainUrl`) when the workspace is not running:

However, there can be multiple reasons why a workspace is not running. Some examples are:
 - workspace is not running because it was shut down due to inactivity
 - workspace is not running because it was shut down due to the run timeout
 - workspace shut down because an error (ie case where for the devworkspace, `status.phase == Failed`)
 - user forgot to start the workspace
 
This PR introduces a new class [WorkspaceStoppedDetector](https://github.com/eclipse-che/che-dashboard/blob/f012219156b24a8c6ba091bd5ae81b2c008d73eb/packages/dashboard-frontend/src/services/bootstrap/workspaceStoppedDetector.ts) to help detect the different cases.
 
Ad a result, different error page messages in each of these four scenarios are available:

<details>
<summary>Workspace inactivity</summary>

![image](https://user-images.githubusercontent.com/83611742/181825169-0b3ea2b6-5d66-40e2-9d48-3c2da6bc8c63.png)

</details>

<details>
<summary>Workspace run timeout</summary>

![image](https://user-images.githubusercontent.com/83611742/181825430-85e82568-81e8-4e31-b02a-e7f1d5a6fa26.png)

</details>

<details>
<summary>Workspace error</summary>

![image](https://user-images.githubusercontent.com/83611742/181825565-bf664356-ea60-45b2-9d9d-633010d2dfcf.png)

</details>

<details>
<summary>Workspace was not started</summary>

![image](https://user-images.githubusercontent.com/83611742/181825481-5049a5c0-060a-43f7-8721-5930a68a0e67.png)

</details>



### What issues does this PR fix or reference?
Fixes https://github.com/eclipse/che/issues/21516


### Is it tested? How?




#### Testing the 'Workspace inactivity' error page
1. Set the workspace inactivity timeout to a smaller value by setting `checluster.spec.devEnvironments.secondsOfInactivityBeforeIdling` to, for example, `120`.

2. Start a workspace. Once the editor opens, let it timeout by not interacting with it for about 120 seconds.

3. Once the workspace times out, refresh the page


#### Testing the 'Workspace run timeout' error page
1. Set the workspace run timeout to a value smaller than the inactivity timeout by setting `checluster.spec.devEnvironments.secondsOfRunBeforeIdling` to, for example, `80`.

2. Start a workspace. After the workspace starts, the workspace should terminate in about 80 seconds.

3. Once the workspace times out, refresh the editor page

#### Testing the 'Workspace error' error page
1. To prevent other timeouts from shutting down the workspace, disable the timeouts by setting both `secondsOfInactivityBeforeIdling` and `secondsOfRunBeforeIdling` to `-1` in the `checluster` CR.

2. Start a workspace with this github repository: https://github.com/dkwon17/failing-workspace/tree/fail-delayed

3. The editor should open, however after roughly 1 minute and 30 seconds, the workspace should terminate due to a failing container.

4. Refresh the editor page

#### Testing the 'Workspace was not started' error page
1. Start a workspace.

2. Once the workspace has started and the editor has loaded, stop the workspace.

3. Refresh the editor page

#### Release Notes

The warning page that appears when accessing the workspace url for a stopped workspace provides a more specific message depending on the reason why the workspace has stopped.

The cases are:
* Workspace has been stopped due to inactivity
* Workspace has been stopped because it reached the run timeout
* Workspace has stopped due to an error causing the workspace pod to shut down
* Workspace was not running in the first place

#### Docs PR
